### PR TITLE
Reneging

### PR DESCRIPTION
**Pull request by geraintpalmer**
Adds reneging feature. Addresses #63 and partially #194.

### CODE REVIEW

File: ciw/import_params.py [hunk 1]
```diff
@@ -72,6 +76,10 @@ def create_network(arrival_distributions=None,
         params['ps_thresholds'] = ps_thresholds
     if server_priority_functions != None:
         params['server_priority_functions'] = server_priority_functions
+    if reneging_time_distributions != None:
+        params['reneging_time_distributions'] = reneging_time_distributions
+    if reneging_destinations != None:
+        params['reneging_destinations'] = reneging_destinations
```
Comment (drvinceknight) [inline ciw/import_params.py:82]:
```suggestion
    if reneging_time_distributions is not None:
        params['reneging_time_distributions'] = reneging_time_distributions
    if reneging_destinations is not None:
        params['reneging_destinations'] = reneging_destinations
```

https://appdividend.com/2022/01/10/python-check-none/

File: ciw/import_params.py [hunk 2]
```diff
@@ -105,6 +113,11 @@ def create_network_from_yml(directory_name):
         for dist in params['service_distributions'][clss]:
             dists.append(get_distribution(dist))
         params['service_distributions'][clss] = dists
+    for clss in params['reneging_time_distributions']:
+        dists = []
+        for dist in params['reneging_time_distributions'][clss]:
+            dists.append(get_distribution(dist))
```
Comment (drvinceknight) [inline ciw/import_params.py:119]:
```suggestion
        dists = [get_distribution(dist) for dist in params['reneging_time_distributions'][clss]]
```

File: ciw/node.py [hunk 3]
```diff
@@ -125,6 +128,9 @@ def begin_service_if_possible_accept(self, next_individual):
             - Update the server's end date (only when servers are not infinite)
         """
         next_individual.arrival_date = self.get_now()
+        if self.reneging:
```
Comment (drvinceknight) [inline ciw/node.py:131]:
```suggestion
        if self.reneging is True:
```

File: ciw/node.py [hunk 4]
```diff
@@ -365,6 +376,8 @@ def have_event(self):
         """
         if self.check_if_shiftchange():
             self.change_shift()
+        elif self.check_if_renege():
```
Comment (drvinceknight) [inline ciw/node.py:379]:
```suggestion
        elif self.check_if_renege() is True:
```

File: ciw/node.py [hunk 5]
```diff
@@ -458,11 +471,41 @@ def release_blocked_individual(self):
...
+        next_node.accept(reneging_individual)
+        self.release_blocked_individual()
+
+
+    def get_reneging_date(self, ind):
+        """
+        Returns a service time for the given customer class.
+        """
+        dist = self.simulation.network.customer_classes[ind.customer_class].reneging_time_distributions[self.id_number-1]
+        if dist == None:
```
Comment (drvinceknight) [inline ciw/node.py:500]:
```suggestion
        if dist is None:
```

File: ciw/node.py [hunk 6]
```diff
@@ -501,20 +544,31 @@ def update_next_event_date(self):
             next individual (who isn't blocked) to end service, or Inf
         """
         next_end_service = float("Inf")
+        next_renege_date = float("Inf")
         if not isinf(self.c):
             for s in self.servers:
                 if s.next_end_service_date < next_end_service:
                     next_end_service = s.next_end_service_date
+            if self.reneging:
```
Comment (drvinceknight) [inline ciw/node.py:552]:
```suggestion
            if self.reneging is True:
```

File: ciw/tests/test_state_tracker.py [hunk 7]
```diff
@@ -1367,4 +1367,67 @@ def test_error_checking_for_state_probabilities(self):
...
+        #### We would expect:
+        # t=0  begin          ---> 0
+        # t=7  arrival cust 1 ---> 1
+        # t=14 arrival cust 2 ---> 2
+        # t=17 renege  cust 2 ---> 1
+        # t=18 leave   cust 1 ---> 0
+        # t=21 arrival cust 3 ---> 1
+        # t=28 arrival cust 4 ---> 2
+        # t=31 renege  cust 4 ---> 1
+        # t=32 leave   cust 3 ---> 0
```
Comment (drvinceknight) [inline ciw/tests/test_state_tracker.py:1391]: I'd suggest moving this to the docstring of `def test_state_tracker_with_reneging` and add a bit more context for clarity.

File: ciw/tests/test_state_tracker.py [hunk 8]
```diff
@@ -1367,4 +1367,67 @@ def test_error_checking_for_state_probabilities(self):
...
+
+        N = ciw.create_network(
+            arrival_distributions=[ciw.dists.Deterministic(7)],
+            service_distributions=[ciw.dists.Deterministic(11)],
+            number_of_servers=[1],
+            reneging_time_distributions=[ciw.dists.Deterministic(3)]
+        )
+        Q = ciw.Simulation(N, tracker=ciw.trackers.NodePopulation())
+        self.assertTrue(Q.nodes[1].reneging)
+        #### We would expect:
```
Comment (drvinceknight) [inline ciw/tests/test_state_tracker.py:1413]: Same moment as above (could break this in to two separate tests).

File: ciw/node.py (hunk 3 above)
Comment (drvinceknight) [reply ciw/node.py:131]: Re that: https://docs.quantifiedcode.com/python-anti-patterns/readability/comparison_to_true.html

File: ciw/node.py [hunk 9]
```diff
@@ -458,11 +471,41 @@ def release_blocked_individual(self):
...
+        self.write_reneging_record(reneging_individual)
+        self.reset_individual_attributes(reneging_individual)
+        self.simulation.statetracker.change_state_renege(self, next_node, reneging_individual, False)
+        next_node.accept(reneging_individual)
+        self.release_blocked_individual()
+
+
+    def get_reneging_date(self, ind):
+        """
+        Returns a service time for the given customer class.
```
Comment (11michalis11) [inline ciw/node.py:497]:
```suggestion
        Returns the reneging date for a given individual.
```

File: docs/Tutorial-I/tutorial_iii.rst [hunk 10]
```diff
@@ -35,6 +35,8 @@ This returns a list of named tuples. Each named tuple contains the following inf
     - :code:`destination`
     - :code:`queue_size_at_arrival`
     - :code:`queue_size_at_departure`
+    - :code:`server_id`
+    - :coee: `record_type`
```
Comment (11michalis11) [inline docs/Tutorial-I/tutorial_iii.rst:39]:
```suggestion
    - :code: `record_type`
```

**Pull request merged**